### PR TITLE
Rchain 3902 - change DeployData to Signed[DeployData] with maximum expirationThreshold

### PR DESCRIPTION
**Pull request by odeac**
## Overview
See title

Expiration threshold is set to maximum for now in order to avoid this PR from getting too big

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3880
https://rchain.atlassian.net/browse/RCHAIN-3902

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: casper/src/test/scala/coop/rchain/casper/ValidateTest.scala [hunk 1]
```diff
@@ -290,9 +287,13 @@ class ValidateTest
   "Future deploy validation" should "work" in withStorage {
     implicit blockStore => implicit blockDagStorage =>
       for {
-        deploy            <- ConstructDeploy.basicProcessedDeploy[Task](0)
-        deployData        = deploy.deploy
-        updatedDeployData = deployData.copy(validAfterBlockNumber = -1)
+        deploy     <- ConstructDeploy.basicProcessedDeploy[Task](0)
+        deployData = deploy.deploy.data
+        updatedDeployData = Signed(
+          deployData.copy(validAfterBlockNumber = -1),
```
Comment (tgrospic) [inline casper/src/test/scala/coop/rchain/casper/ValidateTest.scala:293]: @odeac `-1` should be fine because threshold is set to max, right?

**Review by tgrospic — APPROVED** [review]
LGTM

File: casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala [hunk 2]
```diff
@@ -65,7 +66,7 @@ class MultiParentCasperImpl[F[_]: Sync: Concurrent: Log: Time: SafetyOracle: Las
 
   //TODO: Extract hardcoded version and expirationThreshold
   private val version             = 1L
-  private val expirationThreshold = 50
+  private val expirationThreshold = Int.MaxValue
```
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala:71]: @cboscolo and Rao have confirmed that we are expecting from clients to supply correct block number for `ValidAfterBlockNumber` field in a deploy.

Comment (odeac) [reply casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala:71]: so are you saying that this is not necessary anymore?

Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala:71]: The logic should not change and threshold should stay at 50.

**Review by kayvank — APPROVED** [review]

File: models/src/main/scala/coop/rchain/casper/protocol/CasperMessage.scala [hunk 3]
```diff
@@ -365,39 +371,52 @@ final case class DeployData(
...
-    dd.timestamp,
-    dd.sig,
-    dd.sigAlgorithm,
-    dd.phloPrice,
-    dd.phloLimit,
-    dd.validAfterBlockNumber
-  )
+  implicit val serialize = new Serialize[DeployData] {
+    override def encode(a: DeployData): ByteVector =
+      ByteVector(toProto(a).toByteArray)
```
Comment (tgrospic) [inline models/src/main/scala/coop/rchain/casper/protocol/CasperMessage.scala]: Here is the bug with signing. `DeployData` that is serialized contains the deployer so the signature is always invalid. Before serialization `DeployData` fields that are not part of the signature should be cleared.

File: casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala [hunk 4]
```diff
@@ -185,28 +186,20 @@ class MultiParentCasperImpl[F[_]: Sync: Concurrent: Log: Time: SafetyOracle: Las
...
+      missingUser.asLeft
+    else
+      ().asRight
 
-  def deploy(d: DeployData): F[Either[DeployError, DeployId]] = {
+  def deploy(d: Signed[DeployData]): F[Either[DeployError, DeployId]] = {
     import cats.instances.either._
     import coop.rchain.models.rholang.implicits._
-    validateDeploy(d).fold(
+    validateDeploy(d.data).fold(
```
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala:198]: I don't think this validation is necessary. `DeployData` inside `Signed` type could not be created without the deployer. Am I missing something?

**Review by tgrospic — CHANGES_REQUESTED** [review]

**Review by tgrospic — APPROVED** [review]

**Comment by tgrospic** [comment]
bors try

**Comment by tgrospic** [comment]
@odeac created correct solution to remove the deployer from DeployData. The code compiles but some tests are failing.
https://github.com/odeac/rchain/commits/rchain-3910

File: node/src/main/scala/coop/rchain/node/api/DeployGrpcServiceV1.scala [hunk 5]
```diff
@@ -66,11 +66,18 @@ object DeployGrpcServiceV1 {
...
       def doDeploy(request: DeployDataProto): Task[DeployResponse] =
-        defer(BlockAPI.deploy[F](DeployData.from(request))) { r =>
-          import DeployResponse.Message
-          import DeployResponse.Message._
-          DeployResponse(r.fold[Message](Error, Result))
-        }
+        DeployData
+          .from(request)
+          .fold(
+            errMsg => Task.raiseError[DeployResponse](new Exception(errMsg)),
```
Comment (tgrospic) [inline node/src/main/scala/coop/rchain/node/api/DeployGrpcServiceV1.scala:72]: On error DeployResponse with error should be returned.

**Review by tgrospic — CHANGES_REQUESTED** [review]

**Comment by tgrospic** [comment]
Closing this PR in favor of https://github.com/rchain/rchain/pull/2794 which has complete solution of removing deployer from DeployData without _dirty hack_.

**Pull request closed without merging**